File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/user/UserService.java
package com.pbshop.java.spring.maven.jpa.postgresql.user;

import java.util.Map;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.user.dto.UserDtos.UpdateMeRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.user.dto.UserDtos.UpdateProfileRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.user.dto.UserDtos.UpdateUserStatusRequest;

@Service
@Transactional
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getMe(AuthenticatedUserPrincipal principal) {
        return toUserResponse(requireCurrentUser(principal));
    }

    public Map<String, Object> updateMe(AuthenticatedUserPrincipal principal, UpdateMeRequest request) {
        User user = requireCurrentUser(principal);

        if (request.name() != null && !request.name().isBlank()) {
            user.setName(request.name());
        }
        if (request.phone() != null && !request.phone().isBlank()) {
            user.setPhone(request.phone());
        }
        if (request.password() != null && !request.password().isBlank()) {
            user.setPasswordHash(passwordEncoder.encode(request.password()));
        }

        return toUserResponse(userRepository.save(user));
    }

    public Map<String, Object> deleteMe(AuthenticatedUserPrincipal principal) {
        User user = requireCurrentUser(principal);
        user.setStatus("DELETED");
        userRepository.save(user);

        return Map.of("message", "회원 탈퇴가 완료되었습니다.");
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getProfile(Long userId) {
        return toProfileResponse(findUser(userId));
    }

    public Map<String, Object> updateProfile(AuthenticatedUserPrincipal principal, UpdateProfileRequest request) {
        User user = requireCurrentUser(principal);

        if (request.nickname() != null && !request.nickname().isBlank()) {
            user.setName(request.nickname());
        }
        if (request.bio() != null) {
            user.setBio(request.bio());
        }

        return toProfileResponse(userRepository.save(user));
    }

    public Map<String, Object> updateProfileImage(AuthenticatedUserPrincipal principal, MultipartFile file) {
        User user = requireCurrentUser(principal);
        String originalFilename = file == null || file.getOriginalFilename() == null || file.getOriginalFilename().isBlank()
                ? "profile.png"
                : file.getOriginalFilename();
        user.setProfileImageUrl("https://cdn.pbshop.local/profile/" + user.getId() + "/" + originalFilename);
        userRepository.save(user);

        return Map.of("imageUrl", user.getProfileImageUrl());
    }

    public Map<String, Object> deleteProfileImage(AuthenticatedUserPrincipal principal) {
        User user = requireCurrentUser(principal);
        user.setProfileImageUrl(null);
        userRepository.save(user);

        return Map.of("message", "프로필 이미지가 삭제되었습니다.");
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getUsers(AuthenticatedUserPrincipal principal, int page, int limit, String search) {
        requireAdmin(principal);

        Pageable pageable = PageRequest.of(Math.max(page - 1, 0), Math.max(limit, 1), Sort.by(Sort.Direction.DESC, "id"));
        Page<User> result = (search == null || search.isBlank())
                ? userRepository.findAll(pageable)
                : userRepository.findByEmailContainingIgnoreCaseOrNameContainingIgnoreCase(search, search, pageable);

        return Map.of(
                "items", result.getContent().stream().map(this::toUserResponse).toList(),
                "pagination", Map.of(
                        "page", page,
                        "limit", limit,
                        "total", result.getTotalElements(),
                        "totalPages", result.getTotalPages()
                )
        );
    }

    public Map<String, Object> updateUserStatus(AuthenticatedUserPrincipal principal, Long userId, UpdateUserStatusRequest request) {
        requireAdmin(principal);
        User user = findUser(userId);
        user.setStatus(request.status());
        return toUserResponse(userRepository.save(user));
    }

    private User requireCurrentUser(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }

        return findUser(principal.userId());
    }

    private void requireAdmin(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) {
            throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다.");
        }
    }

    private User findUser(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다."));
    }

    private Map<String, Object> toUserResponse(User user) {
        return Map.of(
                "id", user.getId(),
                "email", user.getEmail(),
                "name", user.getName(),
                "phone", user.getPhone() == null ? "" : user.getPhone(),
                "role", user.getRole(),
                "status", user.getStatus(),
                "emailVerified", user.isEmailVerified(),
                "profileImageUrl", user.getProfileImageUrl() == null ? "" : user.getProfileImageUrl(),
                "bio", user.getBio() == null ? "" : user.getBio()
        );
    }

    private Map<String, Object> toProfileResponse(User user) {
        return Map.of(
                "id", user.getId(),
                "nickname", user.getName(),
                "bio", user.getBio() == null ? "" : user.getBio(),
                "profileImageUrl", user.getProfileImageUrl() == null ? "" : user.getProfileImageUrl()
        );
    }
}

File: BackEnd/TypeScript/typescript-nest-npm-typeorm-postgresql/src/user/user.service.ts
import { Injectable, HttpStatus } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User, UserStatus } from './entities/user.entity';
import { UpdateUserDto } from './dto/update-user.dto';
import { UpdateProfileDto } from './dto/update-profile.dto';
import { BusinessException } from '../common/exceptions/business.exception';
import { hashPassword } from '../common/utils/hash.util';
import { PaginationRequestDto, PaginationResponseDto } from '../common/dto/pagination.dto';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  // ─── USER-01: 내 정보 조회 ───
  async getMe(userId: number) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    return this.toUserResponse(user);
  }

  // ─── USER-02: 내 정보 수정 ───
  async updateMe(userId: number, dto: UpdateUserDto) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    if (dto.name) user.name = dto.name;
    if (dto.phone) user.phone = dto.phone.replace(/[\s-]/g, '');
    if (dto.password) user.password = await hashPassword(dto.password);

    const saved = await this.userRepository.save(user);
    return this.toUserResponse(saved);
  }

  // ─── USER-03: 회원 탈퇴 ───
  async deleteMe(userId: number) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    await this.userRepository.softDelete(userId);
    return { message: '회원 탈퇴가 완료되었습니다.' };
  }

  // ─── USER-04: 회원 목록 조회 (Admin) ───
  async findAll(query: PaginationRequestDto) {
    const [items, totalItems] = await this.userRepository.findAndCount({
      skip: query.skip,
      take: query.limit,
      order: { createdAt: 'DESC' },
    });

    return new PaginationResponseDto(
      items.map((u) => this.toUserResponse(u)),
      totalItems,
      query.page,
      query.limit,
    );
  }

  // ─── USER-05: 회원 상태 변경 (Admin) ───
  async updateStatus(userId: number, status: UserStatus) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    user.status = status;
    const saved = await this.userRepository.save(user);
    return this.toUserResponse(saved);
  }

  // ─── USER-06: 프로필 조회 ───
  async getProfile(userId: number) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    return {
      id: user.id,
      nickname: user.nickname,
      bio: user.bio,
      profileImageUrl: user.profileImageUrl,
      createdAt: user.createdAt,
    };
  }

  // ─── USER-07, 08: 닉네임/소개글 수정 ───
  async updateProfile(userId: number, dto: UpdateProfileDto) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    if (dto.nickname !== undefined) {
      // 닉네임 중복 확인
      const existing = await this.userRepository.findOne({
        where: { nickname: dto.nickname },
      });
      if (existing && existing.id !== userId) {
        throw new BusinessException('USER_NICKNAME_DUPLICATE', HttpStatus.CONFLICT);
      }
      user.nickname = dto.nickname;
    }

    if (dto.bio !== undefined) {
      user.bio = dto.bio || null;
    }

    const saved = await this.userRepository.save(user);
    return {
      id: saved.id,
      nickname: saved.nickname,
      bio: saved.bio,
      profileImageUrl: saved.profileImageUrl,
    };
  }

  // ─── USER-09: 프로필 이미지 관리 ───
  async updateProfileImage(userId: number, imageUrl: string | null) {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new BusinessException('USER_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    user.profileImageUrl = imageUrl;
    const saved = await this.userRepository.save(user);

    return {
      id: saved.id,
      profileImageUrl: saved.profileImageUrl,
    };
  }

  // ─── 헬퍼: 응답 변환 ───
  private toUserResponse(user: User) {
    return {
      id: user.id,
      email: user.email,
      name: user.name,
      phone: user.phone,
      role: user.role,
      status: user.status,
      emailVerified: user.emailVerified,
      nickname: user.nickname,
      bio: user.bio,
      profileImageUrl: user.profileImageUrl,
      point: user.point,
      createdAt: user.createdAt,
    };
  }
}
